internal/olm/operator/registry/index_image.go

```diff
@@ -124,14 +151,306 @@ func (c IndexImageCatalogCreator) CreateCatalog(ctx context.Context, name string
+
+	for i, _ := range entries {
+		nameMap[entries[i].Name] = true
+		if entries[i].Replaces != "" {
+			replacesMap[entries[i].Replaces] = true
+		}
+	}
+
+	// gets the CSV name that does not appear in any replaces field in the entries array
+	for key, _ := range nameMap {
+		if _, present := replacesMap[key]; !present {
+			return key, nil
+		}
+	}
+
+	// this should not be reached since there must be an edge to upgrade from
+	return "", errors.New("no channel head found")
+}
+
+// handleTraditionalUpgrade upgrades an operator that was installed using OLM. Subsequent upgrades will go through the setupFBCupdates function
+func handleTraditionalUpgrade(ctx context.Context, indexImage string, bundleImage string, channelName string) (string, error) {
+	// render the index image
+	render := action.Render{Refs: []string{indexImage}}
+	log.SetOutput(ioutil.Discard)
+	originalDeclCfg, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		return "", err
+	}
+
+	// render the bundle image
+	render = action.Render{Refs: []string{bundleImage}}
+	log.SetOutput(ioutil.Discard)
+	bundleDeclConfig, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if len(bundleDeclConfig.Bundles) != 1 {
+		return "", errors.New("bundle image must have at least one bundle")
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	// search for the specific channel in which the upgrade needs to take place, and upgrade from the channel head
+	for i, _ := range originalDeclCfg.Channels {
+		if originalDeclCfg.Channels[i].Name == channelName && originalDeclCfg.Channels[i].Package == bundleDeclConfig.Bundles[0].Package {
+			// found specific channel
+			channelHead, err := getChannelHead(originalDeclCfg.Channels[i].Entries)
+			if err != nil {
+				return "", err
+			}
+			entry := declarativeconfig.ChannelEntry{
+				Name:     bundleDeclConfig.Bundles[0].Name,
+				Replaces: channelHead,
+			}
+			originalDeclCfg.Channels[i].Entries = append(originalDeclCfg.Channels[i].Entries, entry)
+			break
+		}
+	}
+
+	// add the upgraded bundle to resulting declarative config
+	originalDeclCfg.Bundles = append(originalDeclCfg.Bundles, bundleDeclConfig.Bundles[0])
+
+	// convert declarative config to string
+	content, err := StringifyDeclConfig(originalDeclCfg)
+
+	if err != nil {
+		return "", fmt.Errorf("error converting declarative config to string: %v", err)
+	}
+
+	if content == "" {
+		return "", errors.New("file-Based Catalog contents cannot be empty")
+	}
+
+	if err = ValidateFBC(originalDeclCfg); err != nil {
+		return "", fmt.Errorf("error validating the generated FBC: %v", err)
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	return content, nil
+}
+
+// setupFBCupdates starts the process of upgrading a bundle in an FBC. This function will recreate the FBC that was generated
+// during run bundle and upgrade a specific bundle in the specified channel.
+func setupFBCupdates(c *IndexImageCatalogCreator, ctx context.Context) error {
```

Comment:
I prefer renaming this to `runFBCUpgrade` instead, since we handle one upgrade at a time. Wdyt?

```suggestion
func runFBCupgrade(c *IndexImageCatalogCreator, ctx context.Context) error {
```